test/unittests/UserTest.php

```diff
@@ -12,6 +12,43 @@
  */
 require_once __DIR__ . '/../../vendor/autoload.php';
 use PHPUnit\Framework\TestCase;
+
+/**
+ * Fake User class that has an exact copy of the User::hasLoggedIn method
```

Comment:
This should not use a fake version of the  class. It should use the actual User class.

Reply:
The user class has too many dependencies in the original class. Unless rewriting the user class for unit testing.

Reply:
If you look at the [method](https://github.com/aces/Loris/blob/53712c1c20b799d9194c3d2781206c17b7a3ab0f/php/libraries/User.class.inc#L638), the factory and database are actually declared within this $count definition, which makes it really hard to test without this Fake class. I think this is related to issue #4989 , but this method creates its database very oddly!

Reply:
I think I can fix this by changing how the database is declared. Again, this is related to issue #4989. I will make that change now and see if it works!